pkg/controllers/elasticsearch/actions/scale.go

```diff
@@ -0,0 +1,212 @@
+package actions
+
+import (
+	"fmt"
+
+	apps "k8s.io/api/apps/v1beta1"
+	core "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerror "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/elasticsearch/util"
+	utilapi "github.com/jetstack/navigator/pkg/util/api"
+)
+
+type Scale struct {
+	// The Elasticsearch cluster resource being operated on
+	Cluster *v1alpha1.ElasticsearchCluster
+	// The node pool being scaled
+	NodePool *v1alpha1.ElasticsearchClusterNodePool
+	// Number of replicas to scale to
+	Replicas int32
+}
+
+var _ controllers.Action = &Scale{}
+
+func (c *Scale) Name() string {
+	return "Scale"
+}
+
+func (c *Scale) Message() string {
+	return fmt.Sprintf("Scaled node pool %q to %d replicas", c.NodePool.Name, c.NodePool.Replicas)
+}
+
+// Execute will scale the s.NodePool statefulset to the desired number of
+// replicas. It will refuse to scale if the cluster is not in a 'state to scale'
+// as defined by s.canScaleNodePool.
+func (s *Scale) Execute(state *controllers.State) error {
+	if s.NodePool == nil || s.Cluster == nil {
+		return fmt.Errorf("internal error: nodepool and cluster must be set")
+	}
```

Comment:
I'd rather not have any 'leaf' bits of code throw panics. If the call-site wants to panic as a result of an error, then sure. But this really should be set. I was just trying to be safe here and catch it, but you are correct it should be all or nothing. I'll update other actions to do the same...

Reply:
I can't see any way that  we'd create a scale action with a nil nodepool or nil cluster pointer.
IMO If we're going to do this check it should be done in the `nextAction` function before we create the action rather than having to do it everytime we attempt to use the pointer.
Happy to change it in a followup branch if you agree.

Reply:
This is more just about making our structures safe - imo, anywhere you touch an exported field you should probably check if it's nil. We could 'gate' these fields with a `NewConstructor` as an alternative.